core/src/main/java/io/questdb/griffin/engine/functions/analytic/RankFunctionFactory.java

```diff
@@ -0,0 +1,182 @@
+        private long next = 1;
+
+        @Override
+        public long getLong(Record rec) {
+            return next;
+        }
+
+        @Override
+        public void toTop() {
+            next = 1;
+        }
+
+        @Override
+        public void init(SymbolTableSource symbolTableSource, SqlExecutionContext executionContext) throws SqlException {
+            toTop();
+        }
+    }
+
+    private static class RankFunction extends LongFunction implements ScalarFunction, AnalyticFunction, Reopenable {
+
+        private final static int VAL_MAX_INDEX = 0;
+        private final static int VAL_CURRENT_INDEX = 1;
+        private final static int VAL_OFFSET = 2;
+        private final VirtualRecord partitionByRecord;
+        private final RecordSink partitionBySink;
+        private final Map map;
+        private int columnIndex;
+        private RecordComparator recordComparator;
+
+        public RankFunction(Map map, VirtualRecord partitionByRecord, RecordSink partitionBySink) {
+            this.partitionByRecord = partitionByRecord;
+            this.partitionBySink = partitionBySink;
+            this.map = map;
+        }
+
+        @Override
+        public void close() {
+            Misc.free(map);
+            Misc.freeObjList(partitionByRecord.getFunctions());
+        }
+
+        @Override
+        public long getLong(Record rec) {
+            // not called
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void initRecordComparator(RecordComparatorCompiler recordComparatorCompiler, ArrayColumnTypes chainTypes, IntList order) {
+            this.recordComparator = recordComparatorCompiler.compile(chainTypes, order);
+        }
+
+        @Override
+        public boolean isReadThreadSafe() {
+            return false;
+        }
+
+        @Override
+        public void pass1(Record record, long recordOffset, AnalyticSPI spi) {
+            partitionByRecord.of(record);
+
+            MapKey mapKey = map.withKey();
+            mapKey.put(partitionByRecord, partitionBySink);
+            MapValue mapValue = mapKey.createValue();
+            long maxIndex = 0;
+            if (mapValue.isNew()) {
+                mapValue.putLong(VAL_MAX_INDEX, 0);
+                mapValue.putLong(VAL_CURRENT_INDEX, 0);
+                mapValue.putLong(VAL_OFFSET, 0);
+            } else {
+                maxIndex = mapValue.getLong(VAL_MAX_INDEX);
+            }
+
+            if (recordComparator == null) {
+                // no order or order dismiss
+                Unsafe.getUnsafe().putLong(spi.getAddress(recordOffset, columnIndex), maxIndex + 1);
+            } else {
+                long currentIndex = mapValue.getLong(VAL_CURRENT_INDEX);
+                long offset = mapValue.getLong(VAL_OFFSET);
+                if (currentIndex == 0) {
+                    mapValue.putLong(VAL_CURRENT_INDEX, currentIndex + 1);
```

Comment:
nit: could be simplified to the following:
```suggestion
                    mapValue.putLong(VAL_CURRENT_INDEX, 1);
```